Only update stylesheet when related config option changes

user202729 opened the pull request:
Huge performance improvement with `:config-source`.

Note that further improvement is possible by batching all calls at the end of `:config-source`.

-------------

By the way, (with master branch, not this branch) I once got this result

```
$ time qutebrowser ':repeat 3 config-source ;; quit'
Segmentation fault (core dumped)
```

No idea why that happens, but may be related to doing `config-source` before some Qt objects are ready.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4808)
<!-- Reviewable:end -->

user202729 commented:
Note that the config item may be used in conditional etc. as well, such as

```
            {% if conf.statusbar.position == 'top' %}
```

in qutebrowser/misc/keyhintwidget so it's better to only check for surrounding spaces, not `{{ ... }}`.

user202729 commented:
By the way: in

```
@functools.lru_cache()
def _render_stylesheet(stylesheet: str) -> str:
    """Render the given stylesheet jinja template."""
    with jinja.environment.no_autoescape():
        template = jinja.environment.from_string(stylesheet)  # <-- this line
    return template.render(conf=val)
```

should we compile the template only once, instead of for each config change?

jgkamat commented on qutebrowser/config/config.py:
```diff
@@ -657,6 +666,17 @@ def __init__(self, obj: QObject,
+                                               self._stylesheet))
+
+            # assert that all of the options exists
+            assert all(instance.get_opt(option) for option in options)
```
I feel like this can be a bit dangerous, if for example, we have a string value (not evaluated) " conf.foo.bar ", it will cause a crash here. Why not just add it anyway. If we have a false positive that dosen't really matter (only downside is taking up space in the set but I really don't think that's a worry.

user202729 replied:
It implies that there's an error in the code somewhere else. If it's a string value it should pass a custom value for `option` or parse the stylesheet better.

user202729 replied:
It's also possible to use a library to do this (see for example https://stackoverflow.com/a/26319381)

jgkamat commented on qutebrowser/config/config.py:
```diff
@@ -637,14 +638,22 @@ class StyleSheetObserver(QObject):
-                 stylesheet: typing.Optional[str],
-                 update: bool) -> None:
+                 stylesheet: Optional[str], update: bool,
+                 options: Optional[FrozenSet[str]] = None) -> None:
```
It might be a little cleaner to avoid Optional here and just default to a empty frozenset instead. That way if the update hook runs improperly it won't crash on `for a in None`

user202729 replied:
That's different, None means "parse from the template", empty means "this doesn't depend	on anything although the template may suggest otherwise".

jgkamat commented on qutebrowser/config/config.py:
```diff
@@ -637,14 +638,22 @@ class StyleSheetObserver(QObject):
     """Set the stylesheet on the given object and update it on changes.
 
+    Note that if update is turned on and the stylesheet accesses any config
+    option with a syntax different from 'conf.abc.def' (with surrounding
```
Maybe it would be a bit clearer to use `' conf.abc.def '`  (with spaces)

Actually, are the spaces required in jinja? Why not just search for config values without the spaces?

user202729 replied:
Not very sure about that, just to be more strict because it could as well access something named `myconf` (?). Perhaps I can change it to use `\b`.

After all it should not be too problematic because the stylesheets are controlled by us (and not user defined)

jgkamat commented:
user202729 writes:

> should we compile the template only once, instead of for each config change?

Yah, that sounds like a good idea. You could probably either store it in a 
class variable or use another lru cache wrapper

user202729 commented:
The best way would be to change the type of `STYLESHEET` class variable to the template type. Even if it's parsed on StyleSheetObserver it is still done once per window (which is still better than current solution)

user202729 commented:
Changing the type of `STYLESHEET` from `str` to `qutebrowser.utils.jinja.jinja2.Template` turns out to be somewhat cumbersome, because it's necessary to use `with no_autoescape`.

I tried to add a cache, but it seems to not improve performance after this change. The code isn't much cleaner either. (actually, I think that my config doesn't have any color change) (Try out https://github.com/user202729/qutebrowser/tree/tmp7. Does that make config set faster?)

<sub>(There are some other places that adding a cache may help, such as in `jinja.render` which is used to render `qute://` html pages. Actually they are not used very frequently, but if they're slow then this part can be improved. However doing that will not change the page if the version on disc changes. That shouldn't be too much of a problem)</sub>

The only stylesheet that doesn't use `STYLESHEET` variable is that of `Message` in `qutebrowser/mainwindow/messageview.py`. (by the way the redundant comment below is removed in PR 4490)

The-Compiler commented:
FWIW I'd be much more comfortable with using the [Jinja lowlevel API](http://jinja.pocoo.org/docs/dev/api/#low-level-api) to get that information from the templates, though I'm not sure how well `conf.foo.bar` accesses can be extracted from it.

user202729 commented:
It's also possible to implement our own search algorithm instead of using `find_all`. It turns out that this is shorter.

https://github.com/user202729/qutebrowser/tree/tmp8

The-Compiler commented:
@user202729 I only took a quick look, but I think that's nicer!

user202729 commented:
(Apparently both  `import jinja2` and `import jinja2.nodes` is required for mypy to typecheck properly. In python runtime it makes no difference)

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -640,23 +640,32 @@ class StyleSheetObserver(QObject):
         _stylesheet: The stylesheet template to use.
+        _options: The config options that the stylesheet uses. When it's not
+                  necessary to listen for config changes, this attribute may be
+                  missing.
```
An attribute which is only maybe there is weird - just set it to `None` instead.

user202729 replied:
No opinion on this, but the assignment of None would be redundant because it's never checked for.

The-Compiler replied:
IMHO it's good practice for an object to either have an attribute, or not have it - but not to sometimes have it and sometimes not.

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -640,23 +640,32 @@ class StyleSheetObserver(QObject):
 
+        if update:
+            self._options = jinja.template_config_variables(
+                self._stylesheet)  # type: FrozenSet[str]
```
Is the `type:` annotation needed?

user202729 replied:
Not for understanding code correctness, but that reads like "checking membership for this container is fast" to me.

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -640,23 +640,32 @@ class StyleSheetObserver(QObject):
+            self._options = jinja.template_config_variables(
+                self._stylesheet)  # type: FrozenSet[str]
+
+            # assert that all of the options exists
```
```suggestion
            # assert that all options exist
```

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -640,23 +640,32 @@ class StyleSheetObserver(QObject):
+                self._stylesheet)  # type: FrozenSet[str]
+
+            # assert that all of the options exists
+            assert all(instance.get_opt(option) for option in self._options)
```
This seems a bit weird to me, since it suggests that `get_opt` would return a boolean - but we're actually just checking whether there is no exception, right?

I'd prefer:

```python
for option in self._options:
    instance.get_opt(option)
```

instead. Also, shouldn't this be part of `jinja.template_config_variables` instead? It seems like a good sanity check to have there, and then you can probably also benefit from the caching and not do this multiple times for the same stylesheet.

user202729 replied:
I find it a bit weird too, but I prefer stuffing sanity check into assertions.

Yes, that may be a good idea, the function is specialized for config already.

user202729 replied:
There seems to be quite some (currently 3) of `get_opt` currently used to check existence. Perhaps we should convert them to `assert has_opt(name)`. Also avoid having to read some comment.

user202729 replied:
Is this a good idea?

https://github.com/user202729/qutebrowser/tree/assert-config-has-opt

The-Compiler replied:
Sounds good!

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -665,16 +674,21 @@ def _get_stylesheet(self) -> str:
         self._obj.setStyleSheet(self._get_stylesheet())
 
+    @pyqtSlot(str)
+    def _update_stylesheet_conditional(self, option: str) -> None:
```
For consistency with other similar methods, I'd call this `_maybe_update_stylesheet`.

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -665,16 +674,21 @@ def _get_stylesheet(self) -> str:
         return _render_stylesheet(self._stylesheet)
 
-    @pyqtSlot()
     def _update_stylesheet(self) -> None:
```
Since this is only one line and only called from `_update_stylesheet_conditional` I don't see the benefit of having another method - just call `self._obj.setStyleSheet` in `_update_stylesheet_conditional` directly.

user202729 replied:
May be a good idea. This is private interface and it can be refactor out again if convenient.

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
 js_environment = jinja2.Environment(loader=Loader('javascript'))
+
+
+@functools.lru_cache()
```
Please add this cache to the `:debug-cache-stats` command.

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
+
+
+@functools.lru_cache()
+def template_config_variables(template: str) -> typing.FrozenSet[str]:
```
Please add tests for this in `tests/unit/utils/test_jinja.py` - this should be quite easy to test, and `jinja.py` is marked as one of the files which have 100% test coverage.

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
+@functools.lru_cache()
+def template_config_variables(template: str) -> typing.FrozenSet[str]:
+    """Return a frozenset of config variables that is used in template."""
+    pending = [environment.parse(template)]  # a list of unvisited nodes
```
I don't think the comment adds much - you could even call the variable `unvisited_nodes` instead (but `pending` is fine too, IMHO).

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
+def template_config_variables(template: str) -> typing.FrozenSet[str]:
+    """Return a frozenset of config variables that is used in template."""
+    pending = [environment.parse(template)]  # a list of unvisited nodes
+    result = []  # type: typing.List[str]
```
Why not make this a set and return it directly, instead of making it a list and converting it to a frozenset?

user202729 replied:
May break lru cache if something tamper with the result.

The-Compiler replied:
Fair point, but then this should be a Set instead - there's no point in it being a list, we don't care about duplicates or order.

user202729 replied:
I don't think it matters. We don't need any set functionalities in this case (only add an element and iterate)

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
+            # List of attribute names in reverse order.
+            # For example it's ['ab', 'c', 'd'] for 'conf.d.c.ab'.
+            attrlist = []  # type: typing.List[str]
+            while True:
```
```suggestion
            while isinstance(node, jinja2.nodes.Getattr):
```

makes the while loop clearer, and you can skip the `break` below.

user202729 replied:
I don't think there's any difference about code readability, but your suggested change adds an is_instance check. Not that this is performance critical code, but no reason to sacrifice performance.

The-Compiler replied:
I disagree - I find it much more readable to have the condition in the `while ...:` part (rather than have a `while True:` with a `break`. There's also no reason to sacrifice readability if there isn't an actual performance difference :wink:

user202729 replied:
Blame Python for not having a `do .. while` or `repeat .. until` loop.

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
+            else:
+                pending.append(node)
+        else:
+            pending.extend(node.iter_child_nodes())
```
I think it'd be clearer if you'd move this check to the beginning of the `while` block:

```python
if not isinstance(node, jinja2.nodes.Getattr):
    pending.extend(node.iter_child_nodes())
    continue
```

and then you can dedent the whole thing.

The-Compiler commented on qutebrowser/utils/jinja.py:
```diff
@@ -127,3 +130,33 @@ def render(template, **kwargs):
+
+@functools.lru_cache()
+def template_config_variables(template: str) -> typing.FrozenSet[str]:
+    """Return a frozenset of config variables that is used in template."""
```
```suggestion
    """Return the config variables used in the template."""
```

(the type is already conveyed by the annotations anyways)

user202729 commented:
Some new problems turned up when `has_opt` is used. Some functions *expect* `NoOptionError` to be raised (and catch that) (mostly test. There's only one non-test usage). So I made a `ensure_has_opt` to raise that. Is this ok:

<details>

```diff
diff --git a/qutebrowser/config/config.py b/qutebrowser/config/config.py
index da41773b5..ff4a53d36 100644
--- a/qutebrowser/config/config.py
+++ b/qutebrowser/config/config.py
@@ -335,23 +335,31 @@ class Config(QObject):
                 self._set_value(values.opt, scoped.value,
                                 pattern=scoped.pattern)
 
-    def get_opt(self, name: str) -> 'configdata.Option':
-        """Get a configdata.Option object for the given setting."""
-        try:
-            return configdata.DATA[name]
-        except KeyError:
+    def has_opt(self, name: str) -> bool:
+        """Check if the config has the given option."""
+        return name in configdata.DATA
+
+    def ensure_has_opt(self, name: str):
+        """Ensure that the config has the given option.
+
+        Raise a NoOptionError when it isn't the case. This can be caught and
+        handled correctly. All config access funtions should use this function
+        so that they raise NoOptionError when the name is invalid (that
+        behavior is expected by unit test for config).
+
+        If this should always pass (except for programmer error), assert +
+        has_opt should be used instead.
+        """
+        if name not in configdata.DATA:
             deleted = name in configdata.MIGRATIONS.deleted
             renamed = configdata.MIGRATIONS.renamed.get(name)
-            exception = configexc.NoOptionError(
+            raise configexc.NoOptionError(
                 name, deleted=deleted, renamed=renamed)
-            raise exception from None
-
-    def has_opt(self, name: str) -> bool:
-        """Check if the config has the given option.
 
-        Mainly used in assertions.
-        """
-        return name in configdata.DATA
+    def get_opt(self, name: str) -> 'configdata.Option':
+        """Get a configdata.Option object for the given setting."""
+        self.ensure_has_opt(name)
+        return configdata.DATA[name]
 
     def get(self,
             name: str,
@@ -386,7 +394,7 @@ class Config(QObject):
         Note that the returned values are not watched for mutation.
         If a URL is given, return the value which should be used for that URL.
         """
-        assert self.has_opt(name)
+        self.ensure_has_opt(name)
         value = self._values[name].get_for_url(url, fallback=fallback)
         return self._maybe_copy(value)
 
@@ -399,7 +407,7 @@ class Config(QObject):
         This gets the overridden value for a given pattern, or
         configutils.UNSET if no such override exists.
         """
-        assert self.has_opt(name)
+        self.ensure_has_opt(name)
         value = self._values[name].get_for_pattern(pattern, fallback=False)
         return self._maybe_copy(value)
 
@@ -411,7 +419,7 @@ class Config(QObject):
         Note that it's impossible to get a mutable object for a URL as we
         wouldn't know what pattern to apply.
         """
-        assert self.has_opt(name)
+        self.ensure_has_opt(name)
 
         # If we allow mutation, there is a chance that prior mutations already
         # entered the mutable dictionary and thus further copies are unneeded
```

</details>

The-Compiler commented:
I'd prefer adjusting tests/code to use `has_opt` (or `get_opt` and catch the exception if it actually needs the return value).

user202729 commented:
> I'd prefer adjusting tests/code to use `has_opt` (or `get_opt` and catch the exception if it actually needs the return value).

It's weird to catch `AssertionError`, isn't it? Or do you mean in tests? (but the assertion error don't contain deleted/renamed information)

The-Compiler commented:
It depends on what code you mean exactly. If something actually needs the `NoOptionError` I'd just continue to use `get_opt` probably.

user202729 commented:
How am I supposed to access the fields of jinja2 nodes? The [stub file](https://github.com/python/typeshed/blob/master/third_party/2and3/jinja2/nodes.pyi) doesn't define `name` property for `Name` or `node` property` for `Getattr`.

user202729 commented:
1. The only part of the code that catches `NoOptionError` is

```
    try:
        current = info.config.get_str(optname)
    except configexc.NoOptionError:
        return None
```

which could be easily with a `has_opt` check.

However I can't be sure if it may be catched by some other `except Error` or `except:`.

2. There are code like this in the config system

```
    def test_getattr_invalid(self, container):
        with pytest.raises(configexc.NoOptionError) as excinfo:
            container.tabs.foobar  # pylint: disable=pointless-statement
        assert excinfo.value.option == 'tabs.foobar'
```

but that could be changed to catch `AssertionError` or `Error` instead. The code that tests for the exception info would need to be removed however.

The-Compiler commented on qutebrowser/config/config.py:
```diff
@@ -379,7 +386,7 @@ def get_obj(self,
         If a URL is given, return the value which should be used for that URL.
         """
-        self.get_opt(name)  # To make sure it exists
+        assert self.has_opt(name)
```
This breaks things because it doesn't raise `NoOptionError` anymore, and also assertions can be disabled with `python -O`.

Why not add a `self.ensure_has_opt(name)` (could even be private?) which also sets `deleted`/`renamed`? Basically, move the `except KeyError:` part of `get_opt` into its own function, then that can also be used in `get_opt` itself.

user202729 replied:
I commented about the change break some part of the code at https://github.com/qutebrowser/qutebrowser/pull/4808#issuecomment-497635407 .

About `ensure_has_opt`: okay. It keeps the current interface (raise `NoOptionError` when there's no option, the caller can catch that)

The-Compiler replied:
Ah, I see that's basically what you proposed in https://github.com/qutebrowser/qutebrowser/pull/4808#issuecomment-497615572 - after looking at things in detail, I agree a `ensure_has_opt` would be the right thing to do.

Note that there's more than one (non-test) place handling `NoOptionError`: It inherits `configexc.Error`, and that is handled in a lot of places.

user202729 commented:
Problem: config.instance may be None in tests.

https://ci.appveyor.com/project/qutebrowser/qutebrowser/builds/25748080

The-Compiler commented:
You'll need to add the `config_stub` fixture to tests which need the config.

The-Compiler commented:
Thanks! I opened https://github.com/python/typeshed/issues/3197 for the mypy issue.
